zypp/PublicKey.cc

```diff
@@ -196,6 +199,114 @@ namespace zypp
+  /// \class PublicKeySignatureData::Impl
+  /// \brief  PublicKeySignatureData implementation.
+  ///////////////////////////////////////////////////////////////////
+
+  struct PublicKeySignatureData::Impl
+  {
+    std::string _keyid;
+    std::string _name;
+    Date        _created;
+    Date        _expires;
+
+  public:
+    /** Offer default Impl. */
+    static shared_ptr<Impl> nullimpl();
+    static Target & target(const Pathname & sysRoot = "/")
+    {
+      if ( ! getZYpp()->getTarget() )
+        getZYpp()->initializeTarget( sysRoot );
+      return *getZYpp()->getTarget();
+    }
+
+  private:
+    friend Impl * rwcowClone<Impl>( const Impl * rhs );
+    /** clone for RWCOW_pointer */
+    Impl * clone() const;
+  };
+
+  shared_ptr<zypp::PublicKeySignatureData::Impl> PublicKeySignatureData::Impl::nullimpl()
+  {
+    static shared_ptr<Impl> _nullimpl( new Impl );
+    return _nullimpl;
+  }
+
+  zypp::PublicKeySignatureData::Impl *PublicKeySignatureData::Impl::clone() const
+  {
+    return new Impl( *this );
+  }
+
+  ///////////////////////////////////////////////////////////////////
+  /// class PublicKeySignatureData
+  ///////////////////////////////////////////////////////////////////
+
+  PublicKeySignatureData::PublicKeySignatureData()
+    : _pimpl( Impl::nullimpl() )
+  {}
+
+  PublicKeySignatureData::PublicKeySignatureData(const _gpgme_key_sig *rawKeySignatureData)
+    : _pimpl (new Impl)
+  {
+    _pimpl->_keyid = str::asString(rawKeySignatureData->keyid);
+    _pimpl->_name = str::asString(rawKeySignatureData->uid);
+    _pimpl->_created = zypp::Date(rawKeySignatureData->timestamp);
+    _pimpl->_expires = zypp::Date(rawKeySignatureData->expires);
+  }
+
+  PublicKeySignatureData::~PublicKeySignatureData()
+  {}
+
+  PublicKeySignatureData::operator bool() const
+  { return !_pimpl->_keyid.empty(); }
+
+  std::string PublicKeySignatureData::id() const
+  { return _pimpl->_keyid; }
+
+  std::string PublicKeySignatureData::name() const
+  { return _pimpl->_name; }
+
+  Date PublicKeySignatureData::created() const
+  { return _pimpl->_created; }
+
+  Date PublicKeySignatureData::expires() const
+  { return _pimpl->_expires; }
+
+  bool PublicKeySignatureData::expired() const
+  { return isExpired( _pimpl->_expires ); }
+
+  int PublicKeySignatureData::daysToLive() const
+  { return hasDaysToLive( _pimpl->_expires ); }
+
+  std::string PublicKeySignatureData::asString() const
+  {
+    std::string nameStr;
+    if (!name().empty()) {
+      nameStr = str::Str() << name() << " ";
+    }
+    else {
+      nameStr = "[User ID not found] ";
+    }
+    return str::Str() << nameStr
+                      << id() << " "
+                      << created().printDate()
+                      << " [" << expiresDetail( expires() ) << "]";
+  }
+
+  bool PublicKeySignatureData::inTrustedRing() const
+  {
+    std::list<PublicKey> allKeys = _pimpl->target().rpmDb().pubkeys();
```

Comment:
@mlandres Do you know if this is an appropriate method for accessing the public keys in the rpmdb? Here is the definition of `target` for ease of reference:

```
    static Target & target(const Pathname & sysRoot = "/")
    {
      if ( ! getZYpp()->getTarget() )
        getZYpp()->initializeTarget( sysRoot );
      return *getZYpp()->getTarget();
    }
```

 This implementation is based on code that I saw in some unit tests to get the current target. I am not sure if it is appropriate to use this same logic in the main code.

Reply:
The Target access here is not the right way. Depending on the workflow, we may have to handle keys and repos for the software selection before the target system is ready and mounted (e.g. for bare metal installation). Authority is always our KeyRing and it's synced behind the scenes with the rpmdb.

[KeyRing::isKeyTrusted](https://github.com/openSUSE/libzypp/blob/master/zypp/KeyRing.h#L259) and [KeyRing::isKeyKnown](https://github.com/openSUSE/libzypp/blob/master/zypp/KeyRing.h#L265) should work here; accessed via ` getZYpp()->keyRing()->`.